src/daft-core/src/array/ops/binary.rs

```diff
@@ -0,0 +1,184 @@
+use std::iter;
+
+use common_error::{DaftError, DaftResult};
+
+use crate::{
+    array::ops::as_arrow::AsArrow,
+    datatypes::{BinaryArray, DaftIntegerType, DaftNumericType, DataArray, UInt64Array},
+};
+
+impl BinaryArray {
+    pub fn length(&self) -> DaftResult<UInt64Array> {
+        let self_arrow = self.as_arrow();
+        let offsets = self_arrow.offsets();
+        let arrow_result = arrow2::array::UInt64Array::from_iter(
+            offsets.windows(2).map(|w| Some((w[1] - w[0]) as u64)),
+        )
+        .with_validity(self_arrow.validity().cloned());
+        Ok(UInt64Array::from((self.name(), Box::new(arrow_result))))
+    }
+
+    pub fn binary_concat(&self, other: &Self) -> DaftResult<Self> {
+        let self_arrow = self.as_arrow();
+        let other_arrow = other.as_arrow();
+
+        let arrow_result = if self_arrow.len() == 1 || other_arrow.len() == 1 {
+            // Handle broadcasting case
+            let (longer_arr, shorter_arr, is_self_longer) = if self_arrow.len() > other_arrow.len()
+            {
+                (self_arrow, other_arrow, true)
+            } else {
+                (other_arrow, self_arrow, false)
+            };
+            let shorter_val = shorter_arr.value(0);
+            longer_arr
+                .iter()
+                .map(|val| {
+                    val.map(|val| {
+                        if is_self_longer {
+                            [val, shorter_val].concat()
+                        } else {
+                            [shorter_val, val].concat()
+                        }
+                    })
+                })
+                .collect::<arrow2::array::BinaryArray<i64>>()
+        } else {
+            // Regular case - element-wise concatenation
+            self_arrow
+                .iter()
+                .zip(other_arrow.iter())
+                .map(|(left_val, right_val)| match (left_val, right_val) {
+                    (Some(left), Some(right)) => Some([left, right].concat()),
+                    _ => None,
+                })
+                .collect::<arrow2::array::BinaryArray<i64>>()
+        };
+
+        Ok(Self::from((self.name(), Box::new(arrow_result))))
+    }
+
+    pub fn substr<I, J>(
+        &self,
+        start: &DataArray<I>,
+        length: Option<&DataArray<J>>,
+    ) -> DaftResult<Self>
+    where
+        I: DaftIntegerType,
+        <I as DaftNumericType>::Native: Ord + TryInto<usize>,
+        J: DaftIntegerType,
+        <J as DaftNumericType>::Native: Ord + TryInto<usize>,
+    {
+        let self_arrow = self.as_arrow();
+
+        // Handle broadcasting for start
+        let start_iter = if start.len() == 1 {
```

Comment:
You could simplify the iterator creation by doing something similar to `create_broadcasted_str_iter` in our [utf8 array ops](https://github.com/Eventual-Inc/Daft/blob/main/src/daft-core/src/array/ops/utf8.rs#L41). Take a look at how that is implemented+used, it's also applicable to concat